crypto-sodium/lib/Crypto/Sodium/Salt.hs

```diff
@@ -134,7 +135,6 @@ mkQuoter name convert = QuasiQuoter { quoteExp, quotePat, quoteType, quoteDec }
 --
 -- This function can fail if there are invalid escape sequences.
 parseEscapes :: MonadFail m => String -> m String
-parseEscapes str = case reverse (readP_to_S (many lexChar) str) of
-  (result, ""):_ -> pure result
-  (_, rest):_ -> fail $ "Failed to parse raw bytes: " <> rest
-  [] -> fail $ "Failed to parse raw bytes (no parse): " <> str
+parseEscapes str = case listToMaybe (readP_to_S (many lexChar <* eof) str) of
+  Just (result, "") -> pure result
+  _ -> fail $ "Failed to parse raw bytes (no parse): " <> str
```

Comment:
Uh... there's a bit of a difference between the original implementation and this. The original implementation always showed the first escape sequence that failed to parse in the error message. This one just shows the whole input, which is less helpful. Unfortunately, `ReadP` can't carry error messages (`MonadFail` instance just ignores the argument), so I'm not sure whether we can do better without allowing partial parses.

Reply:
:thinking: How about this?

```haskell
parseEscapes :: MonadFail m => String -> m String
parseEscapes str = case listToMaybe (readP_to_S (many' lexChar) str) of
  Just (result, "") -> pure result
  Just (result, c:_) -> fail $ "Failed to parse raw bytes " <> str <> " at character '" <> [c] <> "'"
  -- the last case actually shouldn't happen, since parser will happily parse zero characters
  _ -> fail $ "Failed to parse raw bytes: " <> str

many' :: ReadP a -> ReadP [a]
many' p = liftM2 (:) p (many' p) <++ pure []
```

`many'` is a greedy parser.

Reply:
I guess I'll make a small PR with this.